Add a registration system for spreadable block compatibility.

gniftygnome opened the pull request:
This PR adds a registration system for spreadable blocks (blocks which extend SpreadableBlock).  Each "type" of spreadable block has its own registry instance, which maps a bare block (f.e. Dirt) to a spreadable block (f.e. Grass or Mycelium).  Two instances are provided for mods extending vanilla grass and mycelium.  Additional registries can be requested for new modded spreadable types.  I believe this is the simplest implementation appropriate for FAPI which implements enough features to be useful and is minimally intrusive.  For example, this provides enough functionality I expect to be able to use it in Terraform API.

Goals:
* Vanilla grass can spread to modded dirt and create the correct modded grass.
* Modded grass can spread to vanilla dirt or other modded dirt and create the correct vanilla or modded grass.
* The same goes for mycelium.
* Additional registries can be easily created for new blocks which extend SpreadableBlock.

This PR does not address decay of spreadable blocks, because that can be easily handled by mods via mixin.

This PR also does not address allowing mods to change the conditions for spread (such as minimum and maximum light levels), although I could imagine this registry optionally storing a reference to an alternative canSpread method to be used in lieu of vanilla's call to the static method in SpreadableBlock (if only they used an instance method instead ... *sigh*).

I have added tests to the test mod and tested using runTestmodClient, and also by modifying Terrestria to use the API in production.

gniftygnome committed:
Add a registration system for spreadable block compatibility.

gniftygnome committed:
Revert IDEA moving things around unnecessarily

gniftygnome committed:
Code style compliance and add tests.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,86 @@
+	 * @param type The registry type Identifier for the desired spreadable block registry
+	 * @return The SpreadableBlockRegistry for the given ID
+	 */
+	static SpreadableBlockRegistry getOrCreateInstance(Identifier type) {
```
Maybe just `get` here? Im not sure the `OrCreateInstance` is really that useful to the API user.

gniftygnome replied:
It creates so the API user can add a custom SpreadableBlock type.  While I don't have a specific need for this myself at the moment, I can imagine a lot of potential uses.  The test cases I provided include creating a spreadable version of Podzol that spreads to Coarse Dirt as an example.

Juuxel replied:
That name does seem too long considering there's no `get` that doesn't have the "or create" functionality.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/SpreadableBlockMixin.java:
```diff
@@ -0,0 +1,73 @@
+
+	@Shadow
+	private static boolean canSpread(BlockState state, WorldView world, BlockPos pos) {
+		return false;
```
Nit: throw in the shadow method

gniftygnome replied:
Done in 11a0801.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/SpreadableBlockMixin.java:
```diff
@@ -0,0 +1,73 @@
+			method = "randomTick",
+			at = @At(
+					value = "INVOKE_ASSIGN",
+					target = "Lnet/minecraft/util/math/BlockPos;add(III)Lnet/minecraft/util/math/BlockPos;",
```
Nit: a comment sayind where its injecting into might be helpful when porting this.

gniftygnome replied:
Done in 11a0801.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	public static @Nullable SpreadableBlockRegistry getInstanceBySpreadable(@Nullable Block block) {
+		if (block == null || !SPREADABLE_TYPE.containsKey(block)) {
```
Why would block ever be null?

gniftygnome replied:
I'm a C programmer.  I have trust issues.  :P
I've changed the methods to require non-null, in 11a0801.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,86 @@
+	/**
+	 * Gets the spreadable block state (if any) for a given bare block state.
+	 *
+	 * @param bareBlockState The bare block state to search the registry for
```
Just reading this is not super clear to me what `The bare block state` means? Would this be dirt as opposed to grass?

gniftygnome replied:
It's Dirt for the two vanilla registries.  It could be anything for any modded registries.  I'll think about how to clarify that a bit.

gniftygnome replied:
In 11a0801 I've reworded a bunch of the API javadocs.  Hopefully this is better now, but let me know.

gniftygnome committed:
Address some review requests.

gniftygnome committed:
Add an accessor for an immutable copy of SPREADABLE_TYPE.

kevinthegreat1 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java:
```diff
@@ -0,0 +1,144 @@
+public final class SpreadableBlockRegistryImpl implements SpreadableBlockRegistry {
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpreadableBlockRegistryImpl.class);
+
+	private static final Map<Block, Identifier> SPREADABLE_TYPE;
```
This seems to only allow a block to be spreadable. What if I only want one block state of a block to be spreadable?

gniftygnome replied:
I was told to keep it really simple, so I did.  Then I was told it has to cover a whole bunch more use cases (and here's another)...  Effectively at least for now this is pretty much a dead PR.  I intend to support all this stuff over in Terraform API where there aren't so many constraints on the implementation and I can already use MixinExtras.

kevinthegreat1 replied:
I understand your frustration, however, once you add something to FAPI, you basically can't make a breaking change at all to ensure compatibility. Therefore, a new API often goes through a lot of deliberation and review. Also, I'm just asking a question about something I noticed.

gniftygnome replied:
I'm not frustrated, though.  I have an alternative avenue for this.  The only way it impacts me is other mod authors will likely use incompatible mixins because FAPI doesn't provide something.  I don't see this as a case of "diligence" on FAPI's end though.  It's more like a lack of consensus or leadership on what is acceptable for inclusion.  If I had written this just to get it into FAPI then it would be frustrating, but I had to do the work anyway for Terraform API.

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	 */
+	Identifier GRASS = new Identifier("grass");
+	/**
+	 * Registry ID for Minecraft Mycelium Block type spreadable blocks.
```
```suggestion
	 * Registry ID for mycelium-type spreadable blocks.
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+@ApiStatus.NonExtendable
+public interface SpreadableBlockRegistry extends Block2ObjectMap<BlockState> {
+	/**
+	 * Registry ID for Minecraft Grass Block type spreadable blocks.
```
```suggestion
	 * Registry ID for grass-type spreadable blocks.
```
No need to say "Minecraft" here, and can be shorter in general

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+import net.fabricmc.fabric.impl.content.registry.SpreadableBlockRegistryImpl;
+
+/**
+ * Registries of Blocks to BlockStates, defining the spreadable BlockState to replace
```
```suggestion
 * Registries of blocks to block states, defining the spreadable block state to replace
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+
+/**
+ * Registries of Blocks to BlockStates, defining the spreadable BlockState to replace
+ * a bare Block with when a particular type of spreadable Block spreads to the bare black.
```
```suggestion
 * a bare block with when a particular type of spreadable block spreads to the bare block.
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+ * Registries of Blocks to BlockStates, defining the spreadable BlockState to replace
+ * a bare Block with when a particular type of spreadable Block spreads to the bare black.
+ *
+ * <p>For example, to get the registry for Mycelium block spread and then register
+ * a modded block pair where Mycelium spreading to MY_DIRT generates MY_MYCELIUM:
```
```suggestion
 * <p>For example, to get the registry for mycelium block spread and then register
 * a modded block pair where mycelium spreading to {@code MY_DIRT} generates {@code MY_MYCELIUM}:
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	 * if none currently exists.
+	 *
+	 * @param type The registry type Identifier for the desired spreadable block registry
+	 * @return The SpreadableBlockRegistry for the given ID
```
```suggestion
	 * @return The {@code SpreadableBlockRegistry} for the given ID
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	/**
+	 * Gets the spreadable block state (if any) for a given bare block state.
+	 *
+	 * <p>For example, the GRASS registry should return the modded grass block state
```
```suggestion
	 * <p>For example, the {@link #GRASS} registry should return the modded grass block state
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	/**
+	 * Gets the spreadable block state (if any) for a given bare block.
+	 *
+	 * <p>For example, the GRASS registry should return the modded grass block
```
```suggestion
	 * <p>For example, the {@link #GRASS} registry should return the modded grass block
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java:
```diff
@@ -0,0 +1,144 @@
+	private static final Map<Identifier, SpreadableBlockRegistry> REGISTRIES;
+
+	static {
+		SPREADABLE_TYPE = new IdentityHashMap<>();
+		REGISTRIES = new HashMap<>();
```
Nit: These can be inline which is a bit more concise and typical in Fabric API

gniftygnome replied:
*shrugs*  There's a static constructor either way; I always prefer to init in the constructor personally but I've changed these back to inline.

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	 * Adds a registry entry to this registry for the given bare block to spreadable block conversion.
+	 *
+	 * <p>For example, a mod adding a custom dirt type would register its modded dirt block and
+	 * the corresponding modded grass block state to the GRASS registry.
```
```suggestion
	 * the corresponding modded grass block state to the {@link #GRASS} registry.
```

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+
+	/**
+	 * Gets an immutable copy of the spreadable block types map, which maps the
+	 * canonical or "primary" block of the type to the registry for the type.
```
Doesn't `SpreadableBlockRegistry.add` map *all* blocks of the type to registry, not just a "primary" block?

https://github.com/FabricMC/fabric/blob/725e336a53e8714456674d2250107f2479cbfc4e/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java#L97-L102

gniftygnome replied:
Yes, `SpreadableBlockRegistry.add` is called to add all the spreadable block transitions to the associated registries.  The types map is basically a map of registry key (which is the primary spreadable block) to registry.  The map is used by the vanilla mixin to implement vanilla-to-mod spread.  Spread from a modded block must be implemented in the block.

You could argue the API implementation is not complete without a `FabricSpreadableBlock` type thing with a default implementation of the mod->any logic.  I wouldn't argue against it; this PR probably should be rewritten anyway, now that MixinExtras is available in the API.  When I have some time in January, I'd like to take another pass at trying to implement some of the more complicated requests (such as allowing mods to set the new block state).

Juuxel replied:
Right. I was just wondering if it's really the "primary" spreadable block if it applies to *all* spreadable blocks using the registry.

Juuxel reviewed (changes requested):
Mostly javadoc notes

gniftygnome committed:
More requested changes from review.

- Various javadoc changes
- SpreadableBlockRegistry.getOrCreateInstance is now getInstance
- When in Rome, do as the Romans say to (or befriend a centurion)

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @return An immutable copy of the spreadable block types map
+	 */
+	static ImmutableMap<Block, Identifier> getSpreadableTypes() {
```
Is there any reason to use indirect `Identifier`s as the values instead of the registry instance directly?

modmuss50 replied:
I wonder if a RegistryKey is the soltuion here? or maybe leave that until we are forced into it?

Juuxel replied:
It's the id of a SpreadableBlockRegistry, not a vanilla registry entry like Block. In other words, just a key of this map:
https://github.com/FabricMC/fabric/blob/fee11564048c76a1273685b0496d559d5c1ce5a4/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java#L41

The instances for each id are constant once they're created (and the spreadable block registries that are in the map have been created), so it'd be simpler to just have the registry itself in the map IMO.

modmuss50 replied:
Ah yeah, sorry I didnt look at the full context. I think this would be a good idea to change.

Syst3ms replied:
I concur with the other comments on this, should probably map to registry instances directly.

modmuss50 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java:
```diff
@@ -0,0 +1,141 @@
+public final class SpreadableBlockRegistryImpl implements SpreadableBlockRegistry {
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpreadableBlockRegistryImpl.class);
+
+	private static final Map<Block, Identifier> SPREADABLE_TYPE = new IdentityHashMap<>();
+	private static final Map<Identifier, SpreadableBlockRegistry> REGISTRIES = new HashMap<>();
```
is there ever a case where they may be multiple `SpreadableBlockRegistry` for a given `Block`. I was looking at fixing it up following Juzz's comments and it seems there might not be. Unless im missing something this could just be a single Map<Block, SpreadableBlockRegistry>, without having any sort of ID at all?

gniftygnome replied:
It took me a while to remember why I did that.  :)  If mod A adds a spreadable block and mod B optionally extends A, I thought it would be nice for mod B to be able to get the registry for that spreadable block without having to import mod A (and guard the reference) to get the block itself.  Using an identifier makes it possible to just create the required identifier in mod B when fetching the registry.

modmuss50 commented:
Going to keep this one in last call for a little longer, I think there are still some design questions to be clarified.

gniftygnome commented:
> Going to keep this one in last call for a little longer, I think there are still some design questions to be clarified.

Yes, I am still not quite to a good point to pick this up again (and as I mentioned, it could probably be improved now that MixinExtras is available, anyway).

modmuss50 commented:
> Yes, I am still not quite to a good point to pick this up again

No worries at all, Im quite happy to get this to the finish line myself, its 99% of the way there.

Syst3ms commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/SpreadableBlockMixin.java:
```diff
@@ -0,0 +1,79 @@
+					target = "Lnet/minecraft/util/math/BlockPos;add(III)Lnet/minecraft/util/math/BlockPos;",
+					shift = At.Shift.BY, by = 2
+			),
+			locals = LocalCapture.CAPTURE_FAILHARD
```
As you mentioned yourself, `@Local` time. That being said, shifts and conditionals are always a bit weird, have you checked that the output is what we want?
